src/components/contextual-menu/contextual-menu.tsx

```diff
@@ -0,0 +1,127 @@
+import { Component, Prop, h, Host, Element, Watch, State } from "@stencil/core";
+
+@Component({
+  tag: "gxg-contextual-menu",
+  styleUrl: "contextual-menu.scss",
+  shadow: true
+})
+export class ContextualMenu {
+  constructor() {
+    this.detectClickOutsideMenu = this.detectClickOutsideMenu.bind(this);
+  }
+
+  /**
+   * The presence of this attribute makes the menu visible
+   */
+  @Prop({ reflect: true }) visible = false;
+
+  private mouseCoordinates = {
+    x: 0,
+    y: 0
+  };
+
+  private contextualMenuSizes = {
+    width: 0,
+    height: 0
+  };
+
+  @State() firstRightClick = true;
+
+  @Element() el: HTMLElement;
+
+  @Watch("visible")
+  watchHandler(newValue: boolean) {
+    if (newValue) {
+      //Get contextualMenu height and width
+      const contextualMenuWidth = (this.contextualMenuSizes.width = this.el.offsetWidth);
+      const contextualMenuHeight = (this.contextualMenuSizes.height = this.el.offsetHeight);
+
+      //Get available height and width from the mouse pointer to the right and bottom side of the document body, respectively.
+      const availableWidth = window.innerWidth - this.mouseCoordinates.x;
+      const availableHeight = window.innerHeight - this.mouseCoordinates.y;
+
+      const widthOverflow = availableWidth < contextualMenuWidth ? true : false;
+      const heightOverflow =
+        availableHeight < contextualMenuHeight ? true : false;
+
+      //Set contextualMenu position
+      this.el.setAttribute(
+        "style",
+        "left:" +
+          this.mouseCoordinates.x +
+          "px; top:" +
+          this.mouseCoordinates.y +
+          "px;"
+      );
+      if (widthOverflow) {
+        this.el.classList.add("width-overflow");
+      } else {
+        this.el.classList.remove("width-overflow");
+      }
+      if (heightOverflow) {
+        this.el.classList.add("height-overflow");
+      } else {
+        this.el.classList.remove("height-overflow");
+      }
```

Comment:
Seems like this should be done in the render method as a response to a change in a prop or state of the component. What do you think?

Reply:
Sorry, I am not getting this one.
This code reacts to a change of the "visible" prop.
I want to get the mouse coordinates and see if the menu has overflow when the menu is about to be visible.

Reply:
I understand, however, here you should just set a state property like, for example, `this.widthOverflow = true` and then, in the `render` method you could do something like this:

```
render() {
  <Host
    class={{
      "width-overflow": this.widthOverflow,
      "height-overflow": this.heightOverflow
    }}
    style={{
      top: `${this.topPosition}px`,
      left: `${this.leftPosition}px`
    }}
   ...
}
```

That way your code will be more declarative and the render method will transform the current state (meaning state and props) into an HTML representation of the component and the rendering logic will be separated from the rest of the logic of the component.

Reply:
Now I get it.
Thank you.